research/conditional/moe_layers/token_choice.py

```diff
@@ -184,6 +183,27 @@ def forward(self, x: torch.Tensor):
 
         return output
 
+    def choose_expert(self, gate_out) -> tuple[torch.Tensor, torch.Tensor]:
+        checkpointing_enabled = (
+            checkpointing.is_in_first_forward() or checkpointing.is_in_second_forward()
+        )
+        if checkpointing_enabled:
+            if checkpointing.is_in_first_forward():
+                with torch.no_grad():
+                    expert_index = torch.argmax(gate_out, dim=1, keepdim=True)
+                    self._checkpointed_expert_index = expert_index
```

Comment:
ape-brain question: why do we use argmax if checkpointing else max?

Reply:
max and argmax return different thins

Reply:
I realise this. Was hoping for a brief overview of how the approach differs in the two cases. Read it myself in the end.